### PR TITLE
Amp include container for IDT2

### LINKED ISSUE #6563: Add AMP Support for IDT2 Includes
**Is your feature request related to a problem? Please describe.**
Currently we render IDT2 content on Canonical pages only, we'd like to add support for rendering these Includes on AMP pages as well.

In order to render IDT2 content on AMP pages, we need to output a responsive flat image of the content with an alt attribute. Metadata for IDT2 images is available on the Smallprox /amp endpoint. Data for 3 image sizes is always provided for any IDT2 Include: small, medium and large. These translate to the following screen sizes (as defined by VJ during the development of Smallprox):
- Small = mobile
- Medium = tablet
- Large = desktop

_Note that we should take notice of 'small', 'medium' or 'large' rather than the specific width and height pixel dimensions, as these may change._

This information (required to create this image for AMP) is now included in the AReS response, based on data retrieved from the Smallprox /amp 'mutator' (following work completed in ARES-1007).

**Describe the solution you'd like**
Our solution should be using data from the ARES payload, we should render images based on that data and depending on the breakpoint information/screen size.

_Note: This should only be done for the AMP implementation, we should not be replacing the Canonical implementation._

If there is no image available in the ARES data for any reason, then we will not render anything for the IDT2 include on AMP.

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

- There is an a11y requirement to ensure the alt text provided with the IDT2 image metadata is announced on screen readers, so this will need an a11y swarm when complete to check this.

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [x] This feature is expected to need manual testing.

**Additional context**
There is no fallback/error message for IDT2 Includes, if it does not render or no image is available in the ARES response for any reason, then we do not show anything.

Example metadata
screenshot-1

### PR DESCRIPTION
**Pull request by rhenshaw56**
Resolves #6563

**Overall change:**
_Adds AMP support for IDT2 includes._

**Code changes:**

- _created `amp/Idt2Amp.jsx` component for idt2 on amp reusing AmpImg from psammead._
- _created `canonical/index.jsx` container and moved original canonical implementation into it._
- _modified `index.jsx` include container to support includes rendering for various include types on both canonical/amp platforms._
- modified includes preprocessor to return a valid image props for an Image container
- _added ares amp data to mundo fixture._

---

- [X] I have assigned myself to this PR and the corresponding issues
- [X] I have added labels to this PR for the relevant pod(s) affected by these changes
- [X] I have assigned this PR to the Simorgh project

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/containers/Include/index.canonical.jsx
Comment (ghost) [inline src/app/containers/Include/index.canonical.jsx]: Could we use `@bbc/psammead-grid` here.

Comment (rhenshaw56) [reply src/app/containers/Include/index.canonical.jsx]: we could but I'd suggest that piece of work should be isolated in an another issue as we'll be doing the same refactor to the canonical implementation

File: data/mundo/cpsAssets/23263889.json
Comment (HarveyPeachey) [inline data/mundo/cpsAssets/23263889.json:401]: The example include looks very skewed, it also seems it's been converted into an image. Is that intentional?

Screenshot 2020-06-02 at 09 01 36

Comment (andrewscfc) [reply data/mundo/cpsAssets/23263889.json:401]: IDT2s on AMP should be converted to images but should not look skewed I would say. I wonder if the correct image is being used for the breakpoint you viewed at @HarveyPeachey ?

Comment (HarveyPeachey) [reply data/mundo/cpsAssets/23263889.json:401]: This was viewed on my desktop so whichever group that is. The image was also skewed when I reduced the window size to a single column layout.

File: src/app/containers/Include/__snapshots__/index.amp.test.jsx.snap
Comment (andrewscfc) [inline src/app/containers/Include/__snapshots__/index.amp.test.jsx.snap]: Should the `src` and `srcset` have all these `undefined` vars in?

File: src/app/containers/Include/__snapshots__/index.test.jsx.snap
Comment (andrewscfc) [inline src/app/containers/Include/__snapshots__/index.test.jsx.snap]: Do we have snapshot tests for the canonical-includes-container and amp-includes-container container respectively, now this test is doing a shallow render the snapshot is capturing a bit less?

File: data/mundo/cpsAssets/23263889.json
Comment (rhenshaw56) [reply data/mundo/cpsAssets/23263889.json:401]: @HarveyPeachey @andrewscfc I also noticed they looked skewed when adjusting between breakpoints however if you adjust to a certain breakpoint and refresh the browser, that goes away for that breakpoint. Wasn't entirely sure what the cause of this was

Comment (HarveyPeachey) [reply data/mundo/cpsAssets/23263889.json:401]: @rhenshaw56 So the skewed image happens on first page load for me. I'm viewing it incognito in chrome. It's also skewed on every gel group breakpoint

File: src/app/containers/Include/index.amp.jsx
Comment (andrewscfc) [inline src/app/containers/Include/index.amp.jsx]:
```suggestion
  const path = href.replace(/https:\/\/smallprox.vj(.*).tools.bbc.co.uk\//, '');
```
I think it might be a bit clearer if we removed the smallprox part of the URL using a regex. Just a bit easier to picture what we are replacing rather than using relying on indexes from a split.

Could we also raise the fact the smallprox domain is in the response with Ares? Ideally this should not be in the data as it exposes an internal endpoint; that would then mean we could remove this code too.

File: src/app/containers/Include/index.amp.test.jsx
Comment (andrewscfc) [inline src/app/containers/Include/index.amp.test.jsx]:
```suggestion
        href: 'https://www.example.com/idt2/cb1a5166-cfbb-4520-bdac-6159299acff6/image/150',
```
- We can use a full domain here to ensure the logic to extract the image relative path from the full path is tested (I've not used smallprox as it ideally should not be publically visible as discussed above)
- We may as well use a representative path as we would expect from Ares

File: data/mundo/cpsAssets/23263889.json
Comment (mykeels) [reply data/mundo/cpsAssets/23263889.json:401]: I noticed that setting the `amp-img`'s `min-width` instead of `width` as it is currently being done fixes the problem by giving the image room to grow, preventing it being skewed

Comment (mykeels) [reply data/mundo/cpsAssets/23263889.json:401]: This might be something we want addressed in a separate issue.

**Comment by andrewscfc** [comment]
I forgot to mention I think this could do with some integration tests to check for an image being rendered as part of a full story page - we may have to change the test assets to ones using includes though.

File: src/app/containers/Include/__snapshots__/index.amp.test.jsx.snap
Comment (rhenshaw56) [reply src/app/containers/Include/__snapshots__/index.amp.test.jsx.snap]: it should not. [this](https://github.com/bbc/simorgh/pull/6672#discussion_r433741924) should fix this

File: src/app/containers/Include/index.amp.jsx
Comment (rhenshaw56) [reply src/app/containers/Include/index.amp.jsx]: > I think it might be a bit clearer if we removed the smallprox part of the URL using a regex. Just a bit easier to picture what we are replacing rather than using relying on indexes from a split.

@andrewscfc this would mean using the smallprox url in the test as mock data 

> Could we also raise the fact the smallprox domain is in the response with Ares? Ideally this should not be in the data as it exposes an internal endpoint; that would then mean we could remove this code too.

I definitely agree with this

File: src/app/containers/Include/__snapshots__/index.test.jsx.snap
Comment (rhenshaw56) [reply src/app/containers/Include/__snapshots__/index.test.jsx.snap]: > Do we have snapshot tests for the canonical-includes-container and amp-includes-container container respectively.

Yes we do.

> now this test is doing a shallow render the snapshot is capturing a bit less?

The intent was to ensure the logic in this file has sufficient coverage and it shouldn't be dependent on whatever logic in the canonical and amp containers.

File: data/mundo/cpsAssets/23263889.json
Comment (rhenshaw56) [reply data/mundo/cpsAssets/23263889.json:401]: @andrewscfc @HarveyPeachey @mykeels this should be fixed now. can you please confirm that?

File: src/app/containers/Include/index.amp.jsx
Comment (andrewscfc) [reply src/app/containers/Include/index.amp.jsx]: >  this would mean using the smallprox url in the test as mock data

I didn't think about that, I don't see that as being an issue personally, if it's already in the fixture data it is already visible. Did you get a chance to raise this with ares, we could potentially get this sorted before this PR is merged?

File: src/app/containers/Include/index.test.jsx
Comment (andrewscfc) [inline src/app/containers/Include/index.test.jsx]: Could we use representative urls here too?

File: src/integration/pages/storyPage/ampIncludeTests.js
Comment (andrewscfc) [inline src/integration/pages/storyPage/ampIncludeTests.js]: Comment should be removed

File: src/integration/pages/storyPage/ampIncludeTests.js
Comment (andrewscfc) [inline src/integration/pages/storyPage/ampIncludeTests.js:9]: Lovely simple test, looks good!

File: data/mundo/cpsAssets/23263889.json
Comment (andrewscfc) [inline data/mundo/cpsAssets/23263889.json]: I think @lizcameron said this issue has now been resolved and this value should now be: `/idt2/cb1a5166-cfbb-4520-bdac-6159299acff6/image/350`, should we update the fixture data, it looks like further down you have applied the changes required.

File: src/app/containers/Include/amp/Idt2Amp.jsx
Comment (andrewscfc) [inline src/app/containers/Include/amp/Idt2Amp.jsx]: Nice to see the beginnings of the proposed frontend refactor, gives us a place to add extra code if we find we need more than just an `AmpImg` 🙂

File: src/app/containers/Include/amp/Idt2Amp.test.jsx
Comment (andrewscfc) [inline src/app/containers/Include/amp/Idt2Amp.test.jsx]: Should we have the html in the props for IDT2 on amp, we don't use it in on Amp

File: src/app/containers/Include/amp/Idt2Amp.test.jsx
Comment (andrewscfc) [inline src/app/containers/Include/amp/Idt2Amp.test.jsx]:
```suggestion
import Idt2Amp from './Idt2Amp';
```
Just renaming the export to match the default export of `Idt2Amp.jsx`

File: src/app/containers/Include/index.jsx
Comment (andrewscfc) [inline src/app/containers/Include/index.jsx:20]: Nice! I can see how we can layer in the refactor really easily 😎

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js]: Should we remove this is the pre-processor as we don't need this data now it has been processed into an image configuration?

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.js]: Could we extract this out into it's own module with it's own unit tests? That way we could decrease the number of tests for the include converter and make the pre-processor more managable.

File: src/app/containers/Include/amp/Idt2Amp.jsx
Comment (simonsinclair) [inline src/app/containers/Include/amp/Idt2Amp.jsx]: This is probably legacy now. 😄

File: src/app/containers/Include/amp/Idt2Amp.test.jsx
Comment (simonsinclair) [inline src/app/containers/Include/amp/Idt2Amp.test.jsx]: Here's a bit more legacy code, that we can probably prune.

**Review by andrewscfc — COMMENTED** [review]
Looking pretty good, now the PR is pretty big but we are quite close to having something that works, so maybe we should just live with it to deliver IDT2 images for batch 3. I really like the beginnings towards @simonsinclair 's refactor proposal whilst still delivering something for our timeline.

Two main suggestions I have:
- Extract the `imageBlock` functionality out into its own module with tests and decrease the tests for include converter, the test file is getting pretty a large, indicating how monolithic the include converter is becoming.
- Remove the original ares image data from the block in te pre-processor, once you have built the attributes for the image block, you don't need the original data in the container imo.

File: src/app/containers/Include/amp/Idt2Amp.test.jsx
Comment (simonsinclair) [inline src/app/containers/Include/amp/Idt2Amp.test.jsx]: Is the extra container logic necessary? If we used react-testing-library `render()`, this will yield a container for us. It also handles `act()`.

File: src/app/containers/Include/index.jsx
Comment (simonsinclair) [inline src/app/containers/Include/index.jsx:21]: 👍

**Review by simonsinclair — COMMENTED** [review]
Really nice work, Rowland. This is going to pave the way for our future refactor's.

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js]: Missing description

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js]:
```suggestion
  it('should generate image block with default image set to a medium sizes for an IDT2 include type on amp', () => {
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.test.js]:
```suggestion
  it('should generate image block with default image set to a large size for an IDT2 include type on canonical', () => {
```

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:379]: Should the small image be in the `srcset`? I can't see `350` in here from: 
```
small: {
            href: '/idt2/111-222-333-444-555/image/350',
            width: 700,
            height: 1864,
          },
```

**Review by andrewscfc — COMMENTED** [review]
This is looking great, just one question about the small image in the `srcset`

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js
Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/index.test.js:379]: It shouldn't when it's not on amp, @jimjohnsonrollings stated earlier of using large sized images for canonical

**Review by andrewscfc — APPROVED** [review]
Thanks for addressing my comments over many reviews, really appreciate it

**Review by simonsinclair — APPROVED** [review]
Thanks for addressing my comments, Rowland. This is looking great!

File: src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.js
Comment (andrewscfc) [inline src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.js:6]: Should we add a test for this @rhenshaw56 ? Was this picked up in testing?

Comment (rhenshaw56) [reply src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/include/getImageBlock.js:6]: > Should we add a test for this

Adding coverage for it now

> Was this picked up in testing?

Yes it was, some IDT2 on test don't have the amp image data cc @paruchurisilpa

**Comment by simonsinclair** [comment]
We are waiting for the refactor in this PR to be merged, so https://github.com/bbc/simorgh/issues/6837 and other issues relating to individual Includes can be picked up.

**Comment by paruchurisilpa** [comment]
Looks good to me including Rowland's console errors fix..

* http://localhost:7080/gahuza/23299713.amp?renderer_env=test
* http://localhost:7080/mundo/23263889.amp?renderer_env=test

**Pull request merged**
